### PR TITLE
I/O redesign

### PR DESCRIPTION
**Pull request by willcrichton**

### CODE REVIEW

File: python/scannerpy/column.py [hunk 1]
```diff
@@ -233,22 +233,16 @@ def load(self, ty=None, fn=None, rows=None, workers=16):
...
                 enc_input = sampled_frame
             img = self._db.ops.ImageEncoder(frame=enc_input)
-            output_op = self._db.sinks.Column(columns={'img': img})
-            op_args[output_op] = png_table_name
-            job = Job(op_args=op_args)
-            self._db.run(
-                output_op, [job], force=True, show_progress=False)
-            out_tbl = self._db.table(png_table_name)
-            return out_tbl.column('img').load(fn=readers.image)
+            output = [ScannerStream(self._db, png_table_name)]
```
Comment (apoms) [inline python/scannerpy/column.py]: Can we file an issue to avoid doing this? It's really gross that we explode the video into images to return them to the user.

File: python/scannerpy/database.py [hunk 2]
```diff
@@ -1143,28 +1146,30 @@ def table(self, name: str) -> Table:
...
-                raise ScannerException(
-                    'Table with name {} not found'.format(name))
         elif isinstance(name, int):
+            table_id = name
             if name in self._table_id:
-                table_id = name
                 table_name = db_meta.tables[self._table_id[name]].name
-            if table_id is None:
-                raise ScannerException(
-                    'Table with id {} not found'.format(name))
```
Comment (apoms) [inline python/scannerpy/database.py]: Why did you remove this check?

File: python/scannerpy/database.py [hunk 3]
```diff
@@ -1143,28 +1146,30 @@ def table(self, name: str) -> Table:
...
             raise ScannerException('Invalid table identifier')
 
         table = Table(self, table_name, table_id)
-        if self._prefetch_table_metadata:
+        if self._prefetch_table_metadata and table_id in self._table_descriptor:
             table._descriptor = self._table_descriptor[table_id]
 
         return table
 
+    def sequence(self, name: str) -> Column:
```
Comment (apoms) [inline python/scannerpy/database.py]: Documentation, pls.

File: python/scannerpy/database.py [hunk 4]
```diff
@@ -1283,7 +1288,6 @@ def batch(l, n):
 
     def run(self,
             outputs: Union[Sink, List[Sink]],
-            jobs: Sequence[Job],
```
Comment (apoms) [inline python/scannerpy/database.py]: Update docs,  pls.

File: python/scannerpy/database.py [hunk 5]
```diff
@@ -1495,30 +1459,14 @@ def run(self,
...
-                                op._name))
-                    else:
-                        args = op_to_op_args[op]
+                    args = op_to_op_args[op]
+
+                    sink_args.args = args
+
+                    streams = op._streams
+                    storage = streams[0].storage()
+                    storage.delete(self, [s for s in streams if s.exists()])
```
Comment (apoms) [inline python/scannerpy/database.py]: Force deleting everything?

File: python/scannerpy/storage.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
+from scannerpy.protobufs import protobufs
+import scannerpy.types as scannertypes
+from scannerpy.common import ScannerException
+import os
+import pickle
+
+class NullElement:
```
Comment (apoms) [inline python/scannerpy/storage.py]: Docs, pls.

File: python/scannerpy/storage.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    pass
+
+class Storage:
```
Comment (apoms) [inline python/scannerpy/storage.py]: Docs.

File: python/scannerpy/storage.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        raise NotImplementedError
+
+    def sink(self, db, op, streams):
+        raise NotImplementedError
+
+    def delete(self, db, streams):
+        raise NotImplementedError
+
+
+class StoredStream:
```
Comment (apoms) [inline python/scannerpy/storage.py]: Docs.

File: python/scannerpy/sink.py [hunk 9]
```diff
@@ -93,26 +93,42 @@ def __init__(self, db):
...
+                inputs = []
+                if sink_info.variadic_inputs:
+                    inputs.extend(args)
+                else:
+                    for c in sink_info.input_columns:
+                        val = kwargs.pop(c.name, None)
+                        if val is None:
+                            raise ScannerException(
+                                'sink {} required column {} as input'
+                                .format(name, c.name))
```
Comment (apoms) [inline python/scannerpy/sink.py]: It's a little annoying that this error will pop up when we aren't exposing sinks/sources explicitly in the current interface. Is there a better way to show this error? Or should it never come up if not using sinks/sources explicitly?

File: python/scannerpy/column.py (hunk 1 above)
Comment (willcrichton) [reply python/scannerpy/column.py]: Yeah sure.

File: python/scannerpy/database.py (hunk 2 above)
Comment (willcrichton) [reply python/scannerpy/database.py]: Since `ScannerStream` allows for references to tables that don't exist.

File: python/scannerpy/sink.py (hunk 9 above)
Comment (willcrichton) [reply python/scannerpy/sink.py]: It should only come up if the `Storage` classes are implemented incorrectly.

File: python/scannerpy/database.py (hunk 4 above)
Comment (willcrichton) [reply python/scannerpy/database.py]: Done

File: python/scannerpy/database.py (hunk 5 above)
Comment (willcrichton) [reply python/scannerpy/database.py]: Added a `cache_mode` mechanism. What do you think?

Comment (apoms) [reply python/scannerpy/database.py]: Looks good to me!
